compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -2173,15 +2220,17 @@ pub fn encode_metadata(tcx: TyCtxt<'_>) -> EncodedMetadata {
             // It can be removed if it turns out to cause trouble or be detrimental to performance.
             join(|| prefetch_mir(tcx), || tcx.exported_symbols(LOCAL_CRATE));
         },
-    )
-    .0
+    );
 }
 
-fn encode_metadata_impl(tcx: TyCtxt<'_>) -> EncodedMetadata {
-    let mut encoder = opaque::Encoder::new(vec![]);
+fn encode_metadata_impl(tcx: TyCtxt<'_>, path: &Path) {
+    let mut encoder = opaque::FileEncoder::new(path)
+        .unwrap_or_else(|err| tcx.sess.fatal(&format!("failed to create file encoder: {}", err)));
     encoder.emit_raw_bytes(METADATA_HEADER).unwrap();
 
-    // Will be filled with the root position after encoding everything.
+    // Though we had holded the root position historically in this place, we moved it to the end
```

Comment:
I re-read the comment above, which mentions garbage bytes appearing at the end of the metadata file in archives, which explains why those bytes are here.  In that case, keeping the length here is a better solution.  Sorry for the mis-direction.